File: examples/android_java/GuardLib.kt
// GuardLib.kt — Android/Kotlin JNI wrapper
//
// Place in: app/src/main/java/com/yourapp/security/GuardLib.kt
//
// Add to app/build.gradle:
//   android {
//       ...
//       sourceSets {
//           main { jniLibs.srcDirs = ['src/main/jniLibs'] }
//       }
//   }
//
// Place .so files in:
//   src/main/jniLibs/arm64-v8a/libguardlib.so
//   src/main/jniLibs/armeabi-v7a/libguardlib.so
//   src/main/jniLibs/x86_64/libguardlib.so

package com.yourapp.security

import android.util.Base64
import android.util.Log

object GuardLib {
    private const val TAG = "GuardLib"

    init {
        System.loadLibrary("guardlib")
    }

    // ── Native declarations ────────────────────────────────────────────────────

    private external fun guard_init(): Long
    private external fun guard_destroy(handle: Long)
    private external fun guard_verify_config(
        handle: Long,
        jsonBytes: ByteArray,
        sigBytes: ByteArray
    ): String?
    private external fun guard_validate_certificate(
        handle: Long,
        certDer: ByteArray,
        expectedFp: String
    ): Int
    private external fun guard_last_error(): String?
    private external fun guard_version(): String

    // ── State ──────────────────────────────────────────────────────────────────

    private var handle: Long = 0L

    val version: String get() = guard_version()

    // ── Public API ─────────────────────────────────────────────────────────────

    /**
     * Initialize GuardLib. Call once on app startup.
     *
     * @return Result.success on success, Result.failure with reason on detection.
     * WARNING: If self-integrity check fails, the process aborts (SIGABRT).
     */
    fun initialize(): Result<Unit> {
        val h = guard_init()
        return if (h != 0L) {
            handle = h
            Log.i(TAG, "Initialized. Version: ${guard_version()}")
            Result.success(Unit)
        } else {
            val err = guard_last_error() ?: "Unknown error"
            Log.e(TAG, "Init failed: $err")
            Result.failure(GuardException(err))
        }
    }

    /**
     * Verify signed config and extract api_url + api_fingerprint.
     *
     * @param configJson  Raw config.json string
     * @param sigBase64   Signature file content (base64)
     * @return GuardConfig on success, null on failure
     */
    fun verifyConfig(configJson: String, sigBase64: String): GuardConfig? {
        checkInitialized()

        val jsonBytes = configJson.toByteArray(Charsets.UTF_8)
        val sigBytes = try {
            Base64.decode(sigBase64.trim(), Base64.DEFAULT)
        } catch (e: Exception) {
            Log.e(TAG, "Invalid signature base64: $e")
            return null
        }

        val resultJson = guard_verify_config(handle, jsonBytes, sigBytes)
        if (resultJson == null) {
            Log.e(TAG, "verifyConfig failed: ${guard_last_error()}")
            return null
        }

        // Parse minimal JSON result
        return try {
            val apiUrl = extractJsonString(resultJson, "api_url") ?: return null
            val apiFingerprint = extractJsonString(resultJson, "api_fingerprint") ?: return null
            GuardConfig(apiUrl, apiFingerprint)
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse result: $e")
            null
        }
    }

    /**
     * Validate certificate fingerprint.
     *
     * @param certDer    Raw DER bytes of server certificate
     * @param expectedFp Expected fingerprint (uppercase hex, colon-separated)
     * @return true if valid, false if mismatch or error
     */
    fun validateCertificate(certDer: ByteArray, expectedFp: String): Boolean {
        checkInitialized()
        return when (val r = guard_validate_certificate(handle, certDer, expectedFp)) {
            1    -> true
            0    -> { Log.w(TAG, "Certificate fingerprint mismatch"); false }
            else -> { Log.e(TAG, "validateCertificate error: ${guard_last_error()}"); false }
        }
    }

    /**
     * Create an OkHttpClient with certificate pinning via GuardLib.
     *
     * @param expectedFp Fingerprint from verified config
     */
    fun createPinnedOkHttpClient(expectedFp: String): okhttp3.OkHttpClient {
        checkInitialized()
        return okhttp3.OkHttpClient.Builder()
            .hostnameVerifier { _, _ -> true } // we do our own pinning
            .sslSocketFactory(
                createPinnedSSLSocketFactory(expectedFp),
                createTrustManager(expectedFp)
            )
            .build()
    }

    fun destroy() {
        if (handle != 0L) {
            guard_destroy(handle)
            handle = 0L
        }
    }

    // ── Private ────────────────────────────────────────────────────────────────

    private fun checkInitialized() {
        if (handle == 0L) throw IllegalStateException("GuardLib not initialized. Call initialize() first.")
    }

    private fun extractJsonString(json: String, key: String): String? {
        val pattern = Regex("\"$key\"\\s*:\\s*\"([^\"]*)\"")
        return pattern.find(json)?.groupValues?.get(1)
    }

    private fun createTrustManager(expectedFp: String) = object : javax.net.ssl.X509TrustManager {
        override fun checkClientTrusted(chain: Array<java.security.cert.X509Certificate>, authType: String) {}
        override fun getAcceptedIssuers(): Array<java.security.cert.X509Certificate> = emptyArray()
        override fun checkServerTrusted(chain: Array<java.security.cert.X509Certificate>, authType: String) {
            if (chain.isEmpty()) throw javax.net.ssl.SSLException("Empty certificate chain")
            val cert = chain[0]
            val certDer = cert.encoded
            if (!validateCertificate(certDer, expectedFp)) {
                throw javax.net.ssl.SSLException("Certificate fingerprint validation failed")
            }
        }
    }

    private fun createPinnedSSLSocketFactory(expectedFp: String): javax.net.ssl.SSLSocketFactory {
        val trustManager = createTrustManager(expectedFp)
        val sslContext = javax.net.ssl.SSLContext.getInstance("TLS")
        sslContext.init(null, arrayOf(trustManager), null)
        return sslContext.socketFactory
    }
}

data class GuardConfig(val apiUrl: String, val apiFingerprint: String)
class GuardException(message: String) : Exception(message)

File: examples/ios/GuardLib.swift
// GuardLib.swift — iOS/macOS Swift wrapper for GuardLib
//
// Setup:
//   1. Add libguardlib.a to your Xcode target (Build Phases → Link Binary With Libraries)
//   2. Add guardlib.h to your Bridging Header:
//      #import "guardlib.h"
//   3. Add libguardlib.a and guardlib.h to your project navigator
//
// Note: On iOS, the library must be a static library (.a), not .dylib,
// because App Store does not allow embedded dynamic libraries from non-Apple SDKs.
// Build with: cargo build --release --target aarch64-apple-ios

import Foundation
import Security

// MARK: - GuardLib

public class GuardLib {
    public static let shared = GuardLib()
    private var handle: OpaquePointer?

    private init() {}

    // MARK: - Public API

    /// Initialize GuardLib and run threat detection.
    ///
    /// - Throws: `GuardError.initFailed` if root/Frida detected.
    /// - Warning: If self-integrity fails, the process aborts (SIGABRT).
    public func initialize() throws {
        let h = guard_init()
        guard let h = h else {
            let reason = lastError() ?? "Unknown error"
            throw GuardError.initFailed(reason)
        }
        handle = h
        let version = String(cString: guard_version())
        print("[GuardLib] ✅ Initialized. Version: \(version)")
    }

    /// Verify signed config and return api_url + api_fingerprint.
    ///
    /// - Parameters:
    ///   - configJson: Raw string content of config.json
    ///   - sigBase64: Raw string content of config.sig (base64)
    /// - Returns: GuardConfig on success, nil on failure
    public func verifyConfig(configJson: String, sigBase64: String) -> GuardConfig? {
        guard let h = handle else { fatalError("GuardLib not initialized") }

        guard let jsonData = configJson.data(using: .utf8) else { return nil }

        let trimmedSig = sigBase64.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let sigData = Data(base64Encoded: trimmedSig), sigData.count == 64 else {
            print("[GuardLib] ❌ Invalid signature")
            return nil
        }

        let resultPtr: UnsafeMutablePointer<CChar>? = jsonData.withUnsafeBytes { jsonRaw in
            sigData.withUnsafeBytes { sigRaw in
                guard let jsonPtr = jsonRaw.baseAddress?.assumingMemoryBound(to: UInt8.self),
                      let sigPtr = sigRaw.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
                    return nil
                }
                return guard_verify_config(h, jsonPtr, jsonData.count, sigPtr, 64)
            }
        }

        guard let resultPtr = resultPtr else {
            print("[GuardLib] ❌ verifyConfig failed: \(lastError() ?? "unknown")")
            return nil
        }

        let resultJson = String(cString: resultPtr)
        guard_free_string(resultPtr)

        return parseGuardConfig(from: resultJson)
    }

    /// Validate a server certificate against a known fingerprint.
    ///
    /// - Parameters:
    ///   - secCertificate: The server certificate from URLSession challenge
    ///   - expectedFp: Expected SPKI SHA-256 fingerprint (uppercase, colon-separated)
    /// - Returns: true if valid, false to abort connection
    public func validateCertificate(_ secCertificate: SecCertificate, expectedFp: String) -> Bool {
        guard let h = handle else { fatalError("GuardLib not initialized") }

        let certData = SecCertificateCopyData(secCertificate) as Data

        let result: Int32 = certData.withUnsafeBytes { certRaw in
            guard let certPtr = certRaw.baseAddress?.assumingMemoryBound(to: UInt8.self) else {
                return -1
            }
            return expectedFp.withCString { fpPtr in
                guard_validate_certificate(h, certPtr, certData.count, fpPtr)
            }
        }

        switch result {
        case 1:  return true
        case 0:  print("[GuardLib] ❌ Certificate fingerprint mismatch"); return false
        default: print("[GuardLib] ❌ Internal error: \(lastError() ?? "unknown")"); return false
        }
    }

    /// Use in URLSession delegate to enforce certificate pinning.
    ///
    /// Usage:
    ///   func urlSession(_ session: URLSession,
    ///                   didReceive challenge: URLAuthenticationChallenge,
    ///                   completionHandler: ...) {
    ///     let disposition = GuardLib.shared.handleAuthChallenge(challenge, expectedFp: trustedFp)
    ///     completionHandler(disposition, nil)
    ///   }
    public func handleAuthChallenge(
        _ challenge: URLAuthenticationChallenge,
        expectedFp: String
    ) -> URLSession.AuthChallengeDisposition {
        guard challenge.protectionSpace.authenticationMethod == NSURLAuthenticationMethodServerTrust,
              let serverTrust = challenge.protectionSpace.serverTrust else {
            return .cancelAuthenticationChallenge
        }

        guard let cert = SecTrustGetCertificateAtIndex(serverTrust, 0) else {
            return .cancelAuthenticationChallenge
        }

        return validateCertificate(cert, expectedFp: expectedFp)
            ? .useCredential
            : .cancelAuthenticationChallenge
    }

    public func destroy() {
        if let h = handle {
            guard_destroy(h)
            handle = nil
        }
    }

    // MARK: - Private

    private func lastError() -> String? {
        guard let ptr = guard_last_error() else { return nil }
        let msg = String(cString: ptr)
        guard_free_string(ptr)
        return msg
    }

    private func parseGuardConfig(from json: String) -> GuardConfig? {
        guard let data = json.data(using: .utf8),
              let obj = try? JSONSerialization.jsonObject(with: data) as? [String: String],
              let apiUrl = obj["api_url"],
              let apiFingerprint = obj["api_fingerprint"],
              !apiUrl.isEmpty, !apiFingerprint.isEmpty else {
            return nil
        }
        return GuardConfig(apiUrl: apiUrl, apiFingerprint: apiFingerprint)
    }
}

// MARK: - Models

public struct GuardConfig {
    public let apiUrl: String
    public let apiFingerprint: String
}

public enum GuardError: Error {
    case initFailed(String)
    case notInitialized

    public var localizedDescription: String {
        switch self {
        case .initFailed(let r): return "GuardLib init failed: \(r)"
        case .notInitialized:    return "GuardLib not initialized"
        }
    }
}
